rlog/2024-07-01-bloomfilter.mdx

```diff
@@ -0,0 +1,142 @@
+In particular, false-positives may be an acceptable tradeoff for speed.
+In this blog, we examine Bloom filters.
+Bloom filters are a data structure that can be used for membership proofs.
+
+
+## Introduction
+Everyone is familiar with the process of creating a new account.
+Consider when you enter your desired username.
+Many sites provide real-time feedback, as you type, on the availability of a given string.
+In this scenario, it is necessary that the result is seemingly instant, regardless of the number of existing accounts.
+However, it is not important for the usernames that are flagged as unavailable are, in fact, in use.
+That is, it is sufficient to have a probabilistic check for membership.
+
+**Bloom filters** is a data structure that can be used to accumulate an arbitrary amount of data with a fixed amount of space.
+Specifically, the associated Bloom filter $B$ for a digest of data $D$ can be queried to determine whether an element is (possibly) a member of $D$:
+
+- **0:** The queried element is definitely not a member of digest $D$.
+- **1:** The entry is possibly a member of the digest $D$.
+
+The algorithms associated to Bloom filters, that we will discuss shortly, are deterministic.
+The possibility of false-positives arises from the query algorithm.
+
+### Setup and Algorithms
+A Bloom filter consists of a binary string ${\bf{v}} \in \{0,1\}^n$ and $k$ hash functions $\{h_i: \{0,1\}^* \rightarrow \{0,\dots,n-1\}\}_{i=0}^{k-1}$.
+We note that each hash function $h_i$ is used to determine an index of our binary string ${\bf{v}}$ to flip the associated bit to 1.
+The binary string ${\bf{v}}$ is initialized with every entry as 0.
+The hash functions do not need to be cryptographic hash functions.
+
+
+- **Append:** Suppose that we wish to add the element $x$ to the Bloom filter.
+    - Define the vector ${\bf{b}} \in \{0,\dots,n-1\}^k$ so that ${\bf{b}}[i] := h_i(x)$ for each $i \in \{0,\dots,k-1\}$.
+    - Update the binary string ${\bf{v}}[{\bf{b}}[i]] \leftarrow 1$ for each $i \in \{0,\dots,k-1\}$.
+   
+- **Query:** Suppose that we wish to query the Bloom filter for element $y$.
+    - Return 1 provided ${\bf{v}}[h_i(y)] = 1$ for every $i \in \{0,\dots,k-1\}$. Otherwise, return 0.
+    
+The algorithm **Query** will output 1 for every element $y$ that has been added to the Bloom filter.
+This is a consequence of the **Append** algorithm.
+However, due to potential collisions over a set of hash functions, it is possible for false-positives to occur.
+Moreover, the possibility of collisions make it impossible to remove elements from the Bloom filter.
+
+### Complexity
+The storage of a Bloom filter requires constant space.
+Specifically, the Bloom filter uses $n$ bits regardless of the size of the digest.
+So, regardless of the number of elements that we append, the Bloom filter will use $n$ bits.
+Further, if we assume that each of the $k$ hash functions run in constant time then we can append/query an entry in $O(k)$.
```

Comment:
We probably want to include circuit complexity here too. Since there are multiple hashes, we need to also say that the number of constraints scales linearly with `k`

Reply:
as a consequence of number of constraints, proving time increases

Reply:
The only constraints that need to be checked are $k$ bits. So, I'm not sure this is necessary?

Reply:
I'd imagine the verifying circuit actually performs the hashes to validate that the particular bit is set ~ and therefore, will scale with the number of hashes, yes? might be missing something here

Reply:
Yes. Would scale with $k$. So, that should be $O(k)$? Or are you thinking something more elaborate?

Reply:
$O(k)$ sounds good